File: ios/CoinFactory.swift
//
//  CoinFactory.swift
//  coinswallet
//
//  Created by Tal Grynbaum on 30/06/2023.
//

import Foundation
import WalletCore

class CoinFactory {
  private static var coinMap = [String: (String) -> Coin]()
  
  static func registerCoin(name: String, constructor: @escaping (String) -> Coin) {
    coinMap[name] = constructor
  }
  
  static func createCoin(coinName: String, mnemonic: String) -> Coin {
    guard let constructor = coinMap[coinName.lowercased()] else {
      print("in create coin unsupported", coinName)
      fatalError("Unsupported coin: \(coinName)")
    }
    let coin =  constructor(mnemonic)
    print("after constructor coin: ", coin)
    return coin
  }
  class Coin {
    var wallet: HDWallet
    
    init(mnemonic: String) {
      guard let wallet = HDWallet(mnemonic: mnemonic, passphrase: "") else {
        fatalError("Failed to create HDWallet")
      }
      self.wallet = wallet
    }
    
    func getNewAddress(isTestNet:Bool) -> String {
      fatalError("Subclasses must override getNewAddress()")
    }
    
    func getPrivateKey(isTestNet:Bool) -> String {
      fatalError("Subclasses must override getPrivateKey()")
    }
    
    func signTransaction(rawData: String) -> String {
      fatalError("Subclasses must override signTransaction(rawData:)")
    }
    func getPublicKeyHash() -> String {
      return ""
    }
    func getExtendedPublicKey(isTestNet:Bool) -> String {
      return ""
    }
    func getExtendedPrivateKey(isTestNet:Bool) -> String {
      return ""
    }
    func getDeriveAddresses() -> NSMutableArray {
      return []
    }
    func addCustomDerivation(derivePath:String) -> NSMutableDictionary {
      var params: NSMutableDictionary = [:]
      return params
    }
  }
}

File: android/app/src/main/java/com/coinswallet/CoinFactory.java
// CoinFactory.java
package com.coinswallet;

import wallet.core.jni.*;
import java.util.*;
import java.util.function.Function;
import java.util.function.Supplier;
import android.util.Log;

import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;

public class CoinFactory {

    // private static final Map<String, Supplier<Coin>> coinMap = new HashMap<>();
    private static final Map<String, Function<String, Coin>> coinMap = new HashMap<>();

    public static void registerCoin(String name, Function<String, Coin> constructor) {
        coinMap.put(name, constructor);
    }

    public static Coin createCoin(String coinName, String mnemonic) {
        Function<String, Coin> constructor = coinMap.get(coinName.toLowerCase());
        if (constructor == null) {
            throw new IllegalArgumentException("Unsupported coin: " + coinName);
        }
        return constructor.apply(mnemonic);
    }

    public static abstract class Coin {
        protected HDWallet wallet;

        protected Coin(String mnemonic) {
            this.wallet = new HDWallet(mnemonic, "");
        }

        public static String toHex(byte[] bytes) {
            StringBuilder sb = new StringBuilder();
            for (byte b : bytes) {
                sb.append(String.format("%02X", b));
            }
            return sb.toString();
        }

        public abstract String getNewAddress(Boolean isTestNet);

        public abstract String getPrivateKey(Boolean isTestNet);
        public String getPublicKeyHex(){
            return null;
        };
        public String getExtendedPublicKey(Boolean isTestNet) {return null; };

        public String getExtendedPrivateKey(Boolean isTestNet) {return null; };

        public ReadableArray getDeriveAddresses(){
            return null;
        };
        public ReadableMap addCustomDerivation(String derivePath){
            return null;
        };
        public abstract String signTransaction(String rawData);
    }
}
